include/grpc/event_engine/event_engine.h

```diff
@@ -255,6 +255,41 @@ class EventEngine : public std::enable_shared_from_this<EventEngine> {
     /// values are expected to remain valid for the life of the Endpoint.
     virtual const ResolvedAddress& GetPeerAddress() const = 0;
     virtual const ResolvedAddress& GetLocalAddress() const = 0;
+
+    /// A method which allows users to query whether an Endpoint implementation
+    /// supports a specified extension. The name of the extension is provided
+    /// as an input.
+    ///
+    /// An extension could be any type with a unique string id. Each extension
+    /// may support additional capabilities and if the Endpoint implementation
+    /// supports the queried extension, it should return a valid pointer to the
+    /// extension type.
+    ///
+    /// E.g., use case of an EventEngine::Endpoint supporting a custom
+    /// extension.
+    ///
+    /// class CustomEndpointExtension {
+    ///  public:
+    ///.   static absl::string_view Name() {
+    ///       return "my.namespace.custom_endpoint_extension";
+    ///    }
+    ///    ...
+    /// }
+    ///
+    ///
+    /// class CustomEndpoint :
+    ///        public EventEngine::Endpoint, CustomEndpointExtension {
+    ///   public:
+    ///     void* QueryExtension(absl::string_view id) override {
+    ///       if (id == CustomEndpointExtension::Name()) {
+    ///         return static_cast<CustomEndpointExtension*>(this);
```

Comment:
This method returns a `void*`. The cast is not helping anything here, the caller will have to do the cast.

This API may make more sense returning a bool. For example:

```
if (!endpoint->SupportsExtension(MyExtension::kExtensionName)) {
    // ... do something and ...
    return;
}
auto* my_extension = static_cast<MyExtension*>(endpoint);
```

Reply:
Updated it now.

Reply:
If I don't know the concrete type of endpoint, then how do I know how to `static_cast` to `MyExtension*`?

this seems unsafe

Reply:
suggest adding some tests, especially one that has two extensions and two concrete implementations:

```
class Ext1 {
 public:
  virtual int Foo() = 0;
};

class Ext2 {
 public:
  virtual int Bar() = 0;
};

class EP1 : public ExtendedEndpoint<Ext1, Ext2> {
 public:
  int Foo() { return 1; }
  int Bar() { return 2; }
};
class EP2 : public ExtendedEndpoint<Ext2, Ext1> {
 public:
  int Foo() { return 3; }
  int Bar() { return 4; }
};

TEST: `EP1.Query<Ext1>()->Foo() == 1, EP1.Query<Ext2>()->Bar() == 2, EP2.Query<Ext1>()->Foo() == 3` etc
```

Reply:
Very good point, Craig, my suggestion won't work. Some days I remember how to C++, some days I write documentation.

Reply:
Added a query_extensions_test now and reverted back to the old QueryExtension method which returns a void*